stairway/src/main/java/bio/terra/stairway/FlightFilter.java

```diff
@@ -177,12 +236,180 @@ public FlightFilter submittedTimeSortDirection(
+   * @param op a {@link FlightFilterOp}
+   * @param value some object for comparison
+   * @return A newly created FlightFilterPredicate object
+   * @throws FlightFilterException if predicate is not supplied
+   */
+  public static FlightFilterPredicate.Builder makePredicateInput(
+      String key, FlightFilterOp op, Object value) {
+    return makePredicate(FlightFilterPredicate.FilterType.INPUT, key, op, value, Datatype.STRING);
+  }
+
+  /**
+   * Create a predicate object.
+   *
+   * @param type the type of predicate
+   * @param key name of the parameter to compare
+   * @param op a {@link FlightFilterOp}
+   * @param value some object for comparison
+   * @param datatype comparison datatype for the value
+   * @return A newly created FlightFilterPredicate object
+   */
+  private static FlightFilterPredicate.Builder makePredicate(
+      FlightFilterPredicate.FilterType type,
+      String key,
+      FlightFilterOp op,
+      Object value,
+      Datatype datatype) {
+    if (key == null) {
+      throw new FlightFilterException("Key must be specified in an input filter");
+    }
+    if (value == null && op != FlightFilterOp.EQUAL) {
+      throw new FlightFilterException(
+          "Value cannot be null in an input filter if not doing an equality check");
+    }
+
+    if (op == FlightFilterOp.IN) {
+      return new FlightFilterPredicate.Builder(
+          type, key, op, value, Datatype.LIST, FlightFilter::makeParameterName);
+    } else if (value == null) {
+      return new FlightFilterPredicate.Builder(type, key, op, null, Datatype.NULL, (f) -> null);
+    } else {
+      return new FlightFilterPredicate.Builder(
+          type, key, op, value, datatype, FlightFilter::makeParameterName);
+    }
+  }
+
+  /**
+   * Provide builders for expressions to be ANDed together
+   *
+   * @param expressions Builders for expressions. These are created with static methods on this
+   *     class
+   * @return A newly created expression builder
+   */
+  public static FlightBooleanOperationExpression.Builder makeAnd(
+      FlightFilterPredicateInterface.Builder... expressions) {
+    return new FlightBooleanOperationExpression.Builder(
+        FlightBooleanOperationExpression.Operation.AND, expressions);
+  }
+
+  /**
+   * Provide builders for expressions to be ANDed together
+   *
+   * @param expressions Builders for expressions. These are created with static methods on this
+   *     class
+   * @return A newly created expression builder
+   */
+  public static FlightBooleanOperationExpression.Builder makeOr(
+      FlightFilterPredicateInterface.Builder... expressions) {
+    return new FlightBooleanOperationExpression.Builder(
+        FlightBooleanOperationExpression.Operation.OR, expressions);
+  }
+
   private String makeParameterName() {
     parameterId++;
     return "ff" + parameterId;
   }
 
-  public static class FlightFilterPredicate {
+  public interface FlightFilterPredicateInterface {
+    interface Builder {
```

Comment:
Note that class members like `Builder` aren't inherited so there's no reason or benefit to have this as a member, other than needing to type `FlightFilterPredicateInterface.Builder` instead of `Builder` when you refer to it.

It's also suspicious to me that this is an interface with no members. Looking at the code, I think it's possible to make `storeInputPredicateValue()` and `storeFlightPredicateValue()` both methods on these objects, vs methods in `FlightFilterAccess`. If that works, it should make the code simpler.

Reply:
It's really there to enable backwards compatibility with existing filtering methods.  I want this PR to not be a breaking change for existing implementors and this should allow that

Reply:
I must be missing something, can you provide an example where changing this would break backward compatibility?

Reply:
Yes, FlightFilterPredicate is a public interface (not java interface but an object that a user can instantiate). By turning it into an interface, it makes it so that existing code should be able to be left unchanged

Reply:
I should amend this: by having `FlightFilterPredicateInterface` this allows nesting or/and clauses by having `FlightBooleanOperationExpression` include a list of `FlightFilterPredicateInterface` (which can be `FlightBooleanOperationExpression` or `FlightFilterPredicate` objects) as it's children